File: server/chat-server/src/domain/repository/user.repository.ts
import { User } from '../model/user.js';

export const USER_REPOSITORY = Symbol('UserRepository');

export abstract class UserRepository {
  abstract save(user: User): Promise<User>;
  abstract findById(id: number): Promise<User | null>;
  abstract findByNickname(nickname: string): Promise<User | null>;
  abstract findAllActiveUsers(): Promise<User[]>;
  abstract findByIds(ids: number[]): Promise<Map<number, User>>;
  abstract existsByNickname(nickname: string): Promise<boolean>;
  abstract delete(user: User): Promise<void>;
}

File: server/chat-server-spring/src/main/java/com/netmarble/chat/domain/repository/UserRepository.java
package com.netmarble.chat.domain.repository;

import com.netmarble.chat.domain.model.User;

import java.util.List;
import java.util.Optional;

/**
 * User 도메인 리포지토리 인터페이스
 * Infrastructure Layer에서 구현됨 (DDD 원칙)
 */
public interface UserRepository {
    
    User save(User user);
    
    Optional<User> findById(Long id);
    
    Optional<User> findByNickname(String nickname);
    
    List<User> findAllActiveUsers();
    
    boolean existsByNickname(String nickname);
    
    void delete(User user);
}
